### PR TITLE
Separate creation take profit order when parent pending and executed,…

### PR DESCRIPTION
**Pull request by TymofiiSavinov**
… cancel create take profit order, fix execute take profit order

### CODE REVIEW

**Comment by markblaize** [comment]
Task linked: [CU-861me3bft Update take profit order](https://app.clickup.com/t/861me3bft)

File: contracts/leverage_trading/src/cancel_take_profit_order.rs [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+                    liquidity
+                } else {
+                    panic!("Some problem with liquidity parsing.")
+                }
+            }
+            PromiseResult::Failed => panic!("Ref finance not found liquidity."),
+        };
+
+        let (min_amount_x, min_amount_y) = if parent_order.order_type == OrderType::Long {
+            (U128::from(0), U128::from(liquidity_info.amount.0 - 1000))
```
Comment (vlad-sundunchan) [inline contracts/leverage_trading/src/cancel_take_profit_order.rs:91]:
```
min_amount_y = U128::from(BigDecimal::from(U128(take_profit_info.1.amount)) * (BigDecimal::one() - BigDecimal::from(INACCURACY_RATE)));
```

File: contracts/leverage_trading/src/cancel_take_profit_order.rs [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+        } else {
+            (U128::from(liquidity_info.amount.0 - 1000), U128::from(0))
```
Comment (vlad-sundunchan) [inline contracts/leverage_trading/src/cancel_take_profit_order.rs:93]:
```
min_amount_x = U128::from(BigDecimal::from(U128(take_profit_info.1.amount)) * (BigDecimal::one() - BigDecimal::from(INACCURACY_RATE)));
```

File: contracts/leverage_trading/src/cancel_take_profit_order.rs [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+                                take_profit_order_pair,
+                            ),
+                    );
+            }
+            _ => {}
+        }
+    }
+
+    #[private]
+    pub fn get_take_profit_liquidity_info_callback(
```
[the remark this answers is not in the document]
Comment (TymofiiSavinov) [reply contracts/leverage_trading/src/cancel_take_profit_order.rs:65]: added

File: contracts/leverage_trading/src/cancel_take_profit_order.rs (hunk 1 above)
Comment (TymofiiSavinov) [reply contracts/leverage_trading/src/cancel_take_profit_order.rs]: fixed

File: contracts/leverage_trading/src/cancel_take_profit_order.rs (hunk 2 above)
Comment (TymofiiSavinov) [reply contracts/leverage_trading/src/cancel_take_profit_order.rs]: fixed

File: contracts/leverage_trading/src/cancel_take_profit_order.rs [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+                    .remove_liquidity_from_take_profit_callback(
+                        order_id,
+                        parent_order,
+                        take_profit_info,
+                    ),
+            );
+    }
+
+    #[private]
+    pub fn remove_liquidity_from_take_profit_callback(
```
[the remark this answers is not in the document]
Comment (TymofiiSavinov) [reply contracts/leverage_trading/src/cancel_take_profit_order.rs]: fixed

File: contracts/leverage_trading/src/execute_order.rs [hunk 5]
```diff
@@ -218,7 +222,39 @@ impl Contract {
 
                 [liquidity_info.amount, min_amount_x, min_amount_y]
             }
-            _ => [U128(0); 3_usize], // It is necessary to implement the functionality for order type TP (TakeProfit)
+            OrderType::TakeProfit => {
+                let parent_order = self.get_order_by(order_id.0).unwrap();
+                if parent_order.order_type == OrderType::Long {
+                    let min_amount_x = U128::from(
```
[the remark this answers is not in the document]
Comment (TymofiiSavinov) [reply contracts/leverage_trading/src/execute_order.rs:228]: fixed

File: contracts/leverage_trading/src/execute_order.rs [hunk 6]
```diff
@@ -218,7 +222,39 @@ impl Contract {
...
+
+                    let (sell_token_decimals, _) =
+                        self.view_pair_tokens_decimals(&order.sell_token, &order.buy_token);
+                    let min_amount_x =
+                        self.from_protocol_to_token_decimals(min_amount_x, sell_token_decimals);
+
+                    [liquidity_info.amount, min_amount_x, min_amount_y]
+                } else {
+                    let min_amount_x = U128::from(0);
+                    let min_amount_y = U128::from(
```
[the remark this answers is not in the document]
Comment (TymofiiSavinov) [reply contracts/leverage_trading/src/execute_order.rs:244]: fixed

File: contracts/leverage_trading/src/cancel_take_profit_order.rs [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+                    .with_attached_deposit(NO_DEPOSIT)
+                    .get_liquidity(take_profit_order_pair.1.lpt_id.clone())
+                    .then(
+                        ext_ref_finance::ext(self.ref_finance_account.clone())
+                            .with_unused_gas_weight(1_u64)
+                            .with_attached_deposit(NO_DEPOSIT)
+                            .get_pool(
+                                self.view_pair(&parent_order.sell_token, &parent_order.buy_token)
+                                    .pool_id,
+                            )
```
Comment (vlad-sundunchan) [inline contracts/leverage_trading/src/cancel_take_profit_order.rs:47]: If I'm not mistaken, it won't work. You need to break it down into separate calls and process their results

**Review by vlad-sundunchan — APPROVED** [review]

**Pull request merged**